source/src/microapp/cs_MicroappController.cpp

```diff
@@ -301,391 +288,418 @@ uint8_t* MicroappController::getOutputMicroappBuffer() {
- *
  */
-void MicroappController::softInterrupt() {
-	uint8_t* outputBuffer           = getOutputMicroappBuffer();
-	microapp_cmd_t* outgoingCommand = reinterpret_cast<microapp_cmd_t*>(outputBuffer);
-
-	if (_softInterruptCounter == MAX_CALLBACKS_WITHIN_A_TICK - 1) {
+void MicroappController::generateInterrupt() {
+	// This is probably already checked before this function call, but let's do it anyway to be sure
+	if (!allowInterrupts()) {
+		return;
+	}
+	if (_interruptCounter == MICROAPP_MAX_INTERRUPTS_WITHIN_A_TICK - 1) {
 		LOGv("Last callback (next one in next tick)");
 	}
+	_interruptCounter++;
 
-	if (_softInterruptCounter == MAX_CALLBACKS_WITHIN_A_TICK) {
-		LOGv("Not doing another callback");
-		return;
-	}
-	_softInterruptCounter++;
+	uint8_t* outputBuffer                    = getOutputMicroappBuffer();
+	microapp_sdk_header_t* outgoingInterrupt = reinterpret_cast<microapp_sdk_header_t*>(outputBuffer);
 
-	outgoingCommand->ack = CS_ACK_BLUENET_MICROAPP_REQUEST;
-	bool callAgain       = false;
-	int8_t repeatCounter = 0;
+	// Request an acknowledgement by the microapp indicating status of interrupt
+	outgoingInterrupt->ack                   = CS_ACK_REQUEST;
+	bool callAgain                           = false;
+	bool ignoreRequest                       = false;
+	int8_t repeatCounter                     = 0;
 	do {
-		LOGv("Call [%i,%i,%i], within interrupt %i",
-			 _softInterruptCounter,
-			 repeatCounter,
-			 outgoingCommand->counter,
-			 outgoingCommand->interruptCmd);
+		LOGv("Call [%i,%i], within interrupt", _interruptCounter, repeatCounter);
 		callMicroapp();
-		bool acked = (outgoingCommand->ack == CS_ACK_BLUENET_MICROAPP_REQ_ACK);
-		bool busy  = (outgoingCommand->ack == CS_ACK_BLUENET_MICROAPP_REQ_BUSY);
-		if (acked) {
-			LOGv("Acked interrupt");
-		}
-		if (busy) {
-			LOGv("Not acked, microapp is busy");
+		ignoreRequest = !handleAck();
+		if (ignoreRequest) {
+			return;
 		}
-		callAgain = retrieveCommand();
+		callAgain = handleRequest();
 		repeatCounter++;
 	} while (callAgain);
 }
 
 /*
- * We resume the previously started coroutine.
+ * Do some checks to validate if we want to generate an interrupt for the pin event
+ * and if so, prepare the outgoing buffer and call generateInterrupt()
  */
-void MicroappController::callMicroapp() {
-#if DEVELOPER_OPTION_DISABLE_COROUTINE == 0
-	if (nextCoroutine()) {
+void MicroappController::onGpioUpdate(uint8_t pinIndex) {
+	if (!allowInterrupts()) {
+		LOGi("Interrupts in progress, ignore pin event");
 		return;
 	}
-#else
-	static bool start = false;
-	// just do it from current coroutine context
-	if (!start) {
-		start = true;
-		LOGi("Call address: %p", sharedState->entry);
-		jumpToAddress(sharedState->entry);
-		LOGi("Microapp might have run, but this statement is never reached (no coroutine used).");
+	uint8_t interruptPin = digitalPinToInterrupt(pinIndex);
+	if (!interruptRegistered(CS_MICROAPP_SDK_TYPE_PIN, interruptPin)) {
```

Comment:
I'm not sure if this was already wrong, but this "misses an indirection". We should have an "id" for the soft interrupt handlers in the microapp that represents that handler. Now a "pin" is used which happens to be unique because you can do only one `attachInterrupt` call per pin. However, I would keep this the same for all interrupts, be it pin interrupts, BLE interrupts, etc. For example `if (!interruptRegistered(CS_MICROAPP_SDK_TYPE_BLE, CS_MICROAPP_SDK_BLE_SCAN_SCANNED_DEVICE))` is now limited to one handler. This doesn't scale well. Suppose we want to scan for multiple addresses, or for everything with an RSSI of < 50, then we need a struct to store this information at the bluenet side and subsequently an `id` of the handler that needs the information.

Reply:
For now I renamed the major/minor fields to type/id. Adding a layer of indirection via unique ids can be done in a next PR.